src/daft-csv/src/read.rs

```diff
@@ -54,57 +77,159 @@ async fn read_csv_single(
     include_columns: Option<Vec<&str>>,
     num_rows: Option<usize>,
     has_header: bool,
-    delimiter: Option<u8>,
+    delimiter: u8,
     io_client: Arc<IOClient>,
     io_stats: Option<IOStatsRef>,
+    schema: Option<SchemaRef>,
+    buffer_size: Option<usize>,
+    chunk_size: Option<usize>,
+    max_chunks_in_flight: Option<usize>,
+    estimated_mean_row_size: Option<usize>,
 ) -> DaftResult<Table> {
+    let (schema, estimated_mean_row_size) = match schema {
+        Some(schema) => (schema.to_arrow()?, estimated_mean_row_size),
+        None => {
+            let (schema, estimated_mean_row_size) = read_csv_schema_single(
+                uri,
+                has_header,
+                Some(delimiter),
+                // Read at most 1 MiB when doing schema inference.
+                Some(1 << 20),
+                io_client.clone(),
+                io_stats.clone(),
+            )
+            .await?;
+            (schema.to_arrow()?, Some(estimated_mean_row_size))
+        }
+    };
+    let compression_codec = CompressionCodec::from_uri(uri);
     match io_client
         .single_url_get(uri.to_string(), None, io_stats)
         .await?
     {
         GetResult::File(file) => {
-            read_csv_single_from_reader(
-                File::open(file.path).await?.compat(),
+            read_csv_from_compressed_reader(
+                BufReader::new(File::open(file.path).await?),
+                compression_codec,
                 column_names,
                 include_columns,
                 num_rows,
                 has_header,
                 delimiter,
+                schema,
+                // Default buffer size of 512 KiB.
+                buffer_size.unwrap_or(512 * 1024),
+                // Default chunk size of 64 KiB.
```

Comment:
Do these constants make sense given its a local buffered file?

Reply:
I have a TODO to benchmark locally and tweak these, but I'm assuming that tweaking these won't matter as much for local reads as they do for cloud reads. I can look into tweaking these if you'd like!